common/types.go

```diff
@@ -238,37 +237,170 @@ func init() {
 		var i InternalAddress
 		i.SetBytes(b)
-		return &i
+		return Address{&i}
 	} else {
 		var e ExternalAddress
 		e.SetBytes(b)
-		return &e
+		return Address{&e}
+	}
+}
+
+// SetBytes sets the address to the value of b.
+// If b is larger than len(a), b will be cropped from the left.
+func (a *Address) SetBytes(b []byte) {
+	if BytesToAddress(b).IsInChainScope() {
+		if internal, ok := a.inner.(*InternalAddress); ok { // Note that in the case that they do not match, nothing happens
+			internal.SetBytes(b)
+		} else {
+			fmt.Println("Tried to set an external address to an internal address")
+		}
+	} else {
+		if external, ok := a.inner.(*ExternalAddress); ok {
+			external.SetBytes(b)
+		} else {
+			fmt.Println("Tried to set an internal address to an external address")
+		}
+	}
+}
+
+// Bytes gets the string representation of the underlying address.
+func (a Address) Bytes() []byte { return a.inner.Bytes() }
+
+// Hash converts an address to a hash by left-padding it with zeros.
+func (a Address) Hash() Hash { return a.inner.Hash() }
+
+// Hex returns an EIP55-compliant hex string representation of the address.
+func (a Address) Hex() string {
+	return a.inner.Hex()
+}
+
+// String implements fmt.Stringer.
+func (a Address) String() string {
+	return a.inner.String()
+}
+
+// Format implements fmt.Formatter.
+// Address supports the %v, %s, %v, %x, %X and %d format verbs.
+func (a Address) Format(s fmt.State, c rune) {
+	a.inner.Format(s, c)
+}
+
+// MarshalText returns the hex representation of a.
+func (a Address) MarshalText() ([]byte, error) {
+	return a.inner.MarshalText()
+}
+
+// UnmarshalText parses a hash in hex syntax.
+func (a *Address) UnmarshalText(input []byte) error {
+	var temp InternalAddress
+	hexutil.UnmarshalFixedText("Address", input, temp[:])
+	if temp.IsInChainScope() {
+		if _, ok := a.inner.(*InternalAddress); ok {
+			a.inner = &temp
+		} else {
+			return fmt.Errorf("UnmarshalText: Tried to set an external address to an internal address")
+		}
+	} else {
+		if external, ok := a.inner.(*ExternalAddress); ok {
+			hexutil.UnmarshalFixedText("Address", input, external[:])
+		} else {
+			return fmt.Errorf("UnmarshalText: Tried to set an external address to an internal address")
+		}
+	}
+	return nil
+}
+
+// UnmarshalJSON parses a hash in hex syntax.
+func (a *Address) UnmarshalJSON(input []byte) error {
+	var temp InternalAddress
+	hexutil.UnmarshalFixedJSON(reflect.TypeOf(InternalAddress{}), input, temp[:])
+	if temp.IsInChainScope() {
+		if _, ok := a.inner.(*InternalAddress); ok {
+			a.inner = &temp
+		} else {
+			return fmt.Errorf("UnmarshalJSON: Tried to set an external address to an internal address")
+		}
+	} else {
+		if external, ok := a.inner.(*ExternalAddress); ok {
+			hexutil.UnmarshalFixedJSON(reflect.TypeOf(ExternalAddress{}), input, external[:])
+		} else {
+			return fmt.Errorf("UnmarshalJSON: Tried to set an external address to an internal address")
+		}
 	}
-	return a
+	return nil
```

Comment:
is this return needed?

Reply:
Yes, because the method must return an error or nil (for no error)